addons/format_lame.c

```diff
@@ -0,0 +1,559 @@
+		return -1;
+	}
+
+	ast_debug(3, "LAME decoder input: %d bytes", rres);
+	p->dsamples = mp3lame_hip_decode(p->hgfp, rres, dinput, p->doutput, sample_rate);
+	while (p->dsamples == 0) {
+		p->dsamples = mp3lame_hip_decode(p->hgfp, 0, NULL, p->doutput, sample_rate);
+	}
+	ast_debug(3, "LAME decoder got %d output samples", p->dsamples);
+
+	return p->dsamples * BITS;
+}
+
+static int mp3lame_open(struct ast_filestream *s)
+{
+	struct mp3lame_private *p = s->_private;
+
+	p->encoder_counter = 0;
+	p->decoder_counter = 0;
+
+	return 0;
+}
+
+static void mp3lame_close(struct ast_filestream *s)
+{
+	struct mp3lame_private *p = s->_private;
+
+	if (p->encoder_counter != 0) {
+		mp3lame_encoder_deinit(s);
+	}
+
+	return;
+}
+
+static struct ast_frame *mp3lame_read(struct ast_filestream *s, int *whennext, int frame_size, int sample_rate)
+{
+	int remaining_samples;
+	struct mp3lame_private *p = s->_private;
+
+	if (p->encoder_counter != 0) {
+		ast_log(LOG_ERROR, "MP3 filestream %p is already in encoder mode (lame)", s);
+		return NULL;
+	}
+
+	if (p->decoder_counter == 0) {
+		p->total_bytes_compressed = mp3lame_decoder_init(s);
+		if (p->total_bytes_compressed < 0) {
+			ast_log(LOG_ERROR, "HIP decoder initialization failed %d", p->total_bytes_compressed / BITS);
+			return NULL;
+		}
+		p->total_bytes_decoded = mp3lame_file_read(s, sample_rate);
+		if (p->total_bytes_decoded < 0) {
+			return NULL;
+		}
+		p->doffset = mp3lame_decoder_deinit(p);
+	}
+	p->decoder_counter++;
+
+	AST_FRAME_SET_BUFFER(&s->fr, s->buf, AST_FRIENDLY_OFFSET, frame_size);
+
+	remaining_samples = p->dsamples - (frame_size / BITS);
+	if (remaining_samples >= 0) {
+		/* Output a full ast_frame */
+		s->fr.datalen = frame_size;
+		s->fr.samples = frame_size / BITS;
+		*whennext = frame_size / BITS;
+		memcpy(s->fr.data.ptr, (char *)p->doutput + p->doffset, frame_size);
+		p->doffset += frame_size;
+		p->dsamples = remaining_samples;
+		return &s->fr;
+	} else {
+		/* Output a partial ast_frame */
+		if (p->dsamples > 0) {
+			s->fr.datalen = p->dsamples * BITS;
+			s->fr.samples = p->dsamples;
+			memcpy(s->fr.data.ptr, (char *)p->doutput + p->doffset, p->dsamples * BITS);
+			*whennext = 0;
+			p->dsamples = 0;
+			return &s->fr;
+		} else {
+			p->dsamples = 0;
+			return NULL;
+		}
+	}
+}
+
+static int mp3lame_write(struct ast_filestream *fs, struct ast_frame *f, int sample_rate)
+{
+	int res;
+	int wres;
+	int ret;
+	unsigned char encoder_buffer[sample_rate];
+	struct mp3lame_private *p = fs->_private;
+
+	if (p->decoder_counter != 0) {
+		ast_log(LOG_ERROR, "MP3 filestream %p is already in decoder mode (hip)", fs);
+		return -1;
+	}
+
+	if (p->encoder_counter == 0) {
+		p->sample_rate = sample_rate;
+		ret = mp3lame_encoder_init(p);
+		if (ret < 0) {
+			ast_log(LOG_ERROR, "LAME encoder initialization failed %d", ret);
+			return -1;
+		}
+	}
+	p->encoder_counter++;
+
+	/*
+	* Reinitialize lame encoder before every five minutes of audio data.
```

Comment:
Nitpick, but the `*`s in the comment should line up (the subsequent lines should indent one space)

Also just curious, is this limitation documented somewhere? (maybe include a link?) Has it already been reported to the library owner?
Or did you just discover it on your own?

Reply:
I've discovered it. Only tried on one machine. Did not reported to the library owner. Probably the encoder is designed primarily to encode files, not realtime audio.

Can you check your previous implementation maybe?

Reply:
I mean asterisk -rx "file convert f.slin f.mp3" has worked fine without this reinitialization and with f.slin longer than five minutes.

But when I did a MixMonitor(f.mp3) or Record(f.mp3), the f.mp3 was cut short at exactly five minutes.